### DISCUSSION TITLE
Component rendering order guarantees

### DISCUSSION THREAD

**Discussion by anonimitoraf** [pmndrs/zustand#1663] [Q&A] [answered]
Hi everyone,

We have this pattern in our app:

Top-level component:
```js
function App() {
  const { user } = useStore(store => store.auth)
  if (user) return <Authenticated /> // See below
  else      return <Unauthenticated />
}
```

Page only for authenticated users:
```js
function Authenticated() {
  const user = useUser() // See below
  return <div>{user.name}</div>
}
```

Some hook that relies on `user` being defined:
```js
function useUser() {
  const { user } = useStore(store => store.auth)
  if (!user) throw Error('missing user!')
  return user
}
```

I would've thought that `user` inside `Authenticated` will always be defined BUT we've come across cases where we'd get the `'missing user!'` error, for example, when the user logs out.
Weirder thing is that this only happens "sometimes". Does this suggest non-deterministic-ness?

This seems to suggest that when `useAuth` changes (e.g. `user` gets set to `null`), `Authenticated` receives this change before `App` does, sometimes, resulting in an error.

Is this the expected behaviour? If so, is there a better way to implement it?

Thanks!

**Comment by dai-shi** [comment, marked as answer]
wonder if it's well-known zombie children problem.
Are you using React 18 or pre-18?

**Reply by anonimitoraf** [reply]
Hi @dai-shi, I'm on React 17. I'll read up a bit on `zombie children`

**Reply by anonimitoraf** [reply]
Interesting, sounds plausible. I had a look at similar issues/discussions:
- https://github.com/pmndrs/zustand/discussions/286
- https://github.com/pmndrs/zustand/issues/302

Do you mind explaining why `unstable_batchedUpdates` solves the problem in these cases? And is this relevant for my case too?

**Reply by dai-shi** [reply]
If I understand the problem correctly, if you upgrade to React 18, it will be fixed.
In React 17, you need to use `unstable_batchedUpdates` from react-dom.

**Reply by anonimitoraf** [reply]
Thanks. At the moment upgrading to React 18 isn't feasible for us. I'm just wondering would a `setTimeout` achieve the same thing as `unstable_batchedUpdates`. 

As far as I understand we just need to prevent synchronous updates to the store?

**Reply by dai-shi** [reply]
It's not that easy to understand it.
As far as I know wrapping your action with `unstable_batchedUpdates` is only the solution in React 17.

**Reply by anonimitoraf** [reply]
I see,thanks. I'll test it out tomorrow

Side question: so does this "zombie child" problem never happen with Zustand + React 18?

**Reply by dai-shi** [reply]
AFAIU, React 18 automatically batches updates unless you explicitly avoids it with `flushSync()`, so basically not.
